File: algorithm/peak_index_in_a_mountain_array.py
class Solution:
    def peakIndexInMountainArray(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        if len(A) < 3:
            return
        lo = 0
        hi = len(A) - 1
        while lo < hi:
            mid = lo + (hi - lo) // 2
            if A[mid - 1] < A[mid] > A[mid + 1]:
                return mid
            elif A[mid] >= A[mid - 1]:
                lo = mid + 1
            elif A[mid] >= A[mid + 1]:
                hi = mid - 1


if __name__ == '__main__':
    A = [0, 1, 0]
    print(Solution().peakIndexInMountainArray(A))

File: Ruby/peak-index-in-a-mountain-array.rb
# @param {Integer[]} a
# @return {Integer}
def peak_index_in_mountain_array(a)
  left = 0
  right = a.size
  while left < right
    mid = left + (right - left) / 2
    if a[mid-1] < a[mid] and a[mid] > a[mid+1]
      return mid
    elsif a[mid] >= a[mid-1]
      left = mid+1
    elsif a[mid] >= a[mid+1]
      right = mid
    end
  end
end

p peak_index_in_mountain_array([0,2,1,0])
